### PR TITLE
secure istio components

### PR DESCRIPTION
**Pull request by rkpagadala**
**What this PR does / why we need it**:
Code for securing istio components

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#1007 

**Special notes for your reviewer**:
This PR goes with changes in API  https://github.com/istio/api/pull/209

**Release note**:

```release-note
Support mutual TLS for istio components pilot and mixer
```

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -463,8 +463,9 @@ go_repository(
 )
 
 # This SHA is obtained from istio/api
-ISTIO_API = "26e98ec08eeaa97ed0bbcdc28fabd79199256073"
-
+#ISTIO_API = "26e98ec08eeaa97ed0bbcdc28fabd79199256073"
+ISTIO_API = "2e6abc1cfc19b5b7dea7a36c8f59ff41381aeead"
```
Comment (rkpagadala) [inline WORKSPACE]: Changes in this file wont be commited, waiting for https://github.com/istio/api/pull/209

File: model/validation.go [hunk 2]
```diff
@@ -1136,5 +1136,12 @@ func ValidateProxyConfig(config *proxyconfig.ProxyConfig) (errs error) {
 		errs = multierror.Append(errs, multierror.Prefix(err, "invalid proxy admin port:"))
 	}
 
+	switch config.ControlPlaneAuthPolicy {
+	case proxyconfig.AuthenticationPolicy_NONE, proxyconfig.AuthenticationPolicy_MUTUAL_TLS:
+	default:
+		errs = multierror.Append(errs,
+			fmt.Errorf("unrecognized control plane auth policy %q", config.ControlPlaneAuthPolicy))
+	}
+
```
Comment (rshriram) [inline model/validation.go]: you don't need this. Enums are validated.

File: model/validation_test.go [hunk 3]
```diff
@@ -851,6 +851,7 @@ func TestValidateProxyConfig(t *testing.T) {
 		ServiceCluster:         "",
 		StatsdUdpAddress:       "10.0.0.100",
 		ZipkinAddress:          "10.0.0.100",
+		ControlPlaneAuthPolicy:        -1,
```
Comment (rshriram) [inline model/validation_test.go]: remove

File: docker/envoy_mixer.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        "lb_type": "round_robin",
+        "hosts": [
+          {
+            "url": "tcp://127.0.0.1:9091"
+          }
+        ]
+      }
+    ]
+  }
+}
```
Comment (rshriram) [inline docker/envoy_mixer.json]: please add the statsd and tracing configuration as well. The proxy agent has CLI options to specify the statsd cluster and the zipkin cluster. Statsd config is useful immediately. Tracing will become useful soon after we get tracing for xDS

File: docker/envoy_mixer.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  "listeners": [
+   {
+    "address": "tcp://0.0.0.0:15004",
+    "name": "tcp_0.0.0.0_15004",
+    "filters": [
+     {
+      "type": "read",
+      "name": "tcp_proxy",
+      "config": {
+       "stat_prefix": "tcp",
```
Comment (rshriram) [inline docker/envoy_mixer.json]: why is this just tcp proxy? we should have http_connection_manager here. Or is the goal to just secure the channel?

File: proxy/context.go [hunk 6]
```diff
@@ -124,6 +127,12 @@ const (
 
 	// IngressKeyFilename is the ingress private key file name
 	IngressKeyFilename = "tls.key"
+
+	// DiscoveryAppListenPort is the port on which the Discovery service listens
+	DiscoveryAppListenPort = 8080
```
Comment (rshriram) [inline proxy/context.go]: PilotListenPort

File: proxy/context.go [hunk 7]
```diff
@@ -124,6 +127,12 @@ const (
...
+
+	// MixerAppListenPort is the port on which the Mixer service listens
+	MixerAppListenPort = 9091
```
Comment (rshriram) [inline proxy/context.go]: MixerPort

File: docker/envoy_mixer.json (hunk 4 above)
Comment (rkpagadala) [reply docker/envoy_mixer.json]: Shriram, we are not using pilot agent to generate the envoy configuration. I could add a static configuration, but it will not work if mesh config is changed. This is why I had pilot-agent generating envoy.json startup file in my previous iteration. 

https://github.com/istio/pilot/pull/1444

File: docker/envoy_mixer.json (hunk 5 above)
Comment (rkpagadala) [reply docker/envoy_mixer.json]: For now the goal is to secure the channel. Will send out a  different PR for getting it to http connection manager.

File: docker/envoy_mixer.json (hunk 4 above)
Comment (rkpagadala) [reply docker/envoy_mixer.json]: "proxy agent has CLI options to specify the statsd cluster and the zipkin cluster"
specifically I can't use the CLI options to proxy agent because the files are static. I prefer my previous approach where proxy agent generates the file... but you vetoed it :)

Comment (rshriram) [reply docker/envoy_mixer.json]: Ah got it. Sorry. ignore the comment

File: cmd/pilot-agent/main.go [hunk 8]
```diff
@@ -221,6 +251,10 @@ func init() {
 		"IP Address and Port of a statsd UDP listener (e.g. 10.75.241.127:9125)")
 	proxyCmd.PersistentFlags().IntVar(&proxyAdminPort, "proxyAdminPort", int(values.ProxyAdminPort),
 		"Port on which Envoy should listen for administrative commands")
+	proxyCmd.PersistentFlags().IntVar(&controlPlaneAuthPolicy, "controlPlaneAuthPolicy",
+		int(values.ControlPlaneAuthPolicy), "Enum value of controlPlaneAuthPolicy")
```
Comment (rshriram) [inline cmd/pilot-agent/main.go]: okay, now I see the disconnect. For clarity, I suggest changing this to a text based option list, like the service registries.. enable, disable. Or even simply, --enable-control-plane-auth. you dont have to block on updating api..

File: platform/kube/inject/testdata/auth.yaml.injected [hunk 9]
```diff
@@ -52,6 +52,8 @@ spec:
         - ""
         - --proxyAdminPort
         - "15000"
+        - --controlPlaneAuthPolicy
+        - "0"
```
Comment (rshriram) [inline platform/kube/inject/testdata/auth.yaml.injected]: yeah. definitely move to either simple --enable-con... or use text strings.

File: cmd/pilot-agent/main.go [hunk 10]
```diff
@@ -119,6 +125,30 @@ var (
...
+						// namespace of pilot is not part of discovery address use pod namespace
+						// e.g. istio-pilot:15003
+						ns = os.Getenv("POD_NAMESPACE")
+					} else {
+						// namespace is found in the discovery address
+						// e.g. istio-pilot.istio-system:15003
+						ns = parts[1]
+					}
+				}
+				pilotSAN = envoy.GetPilotSAN(pilotDomain, ns)
```
Comment (rshriram) [inline cmd/pilot-agent/main.go]: silly question: isnt this common to all proxies?

Comment (rkpagadala) [reply cmd/pilot-agent/main.go]: No. ingress and egress have istio-pilot:15003, whereas apps have istio-pilot.istio-system:15003
where namespace of istio components if istio-system.

File: platform/kube/inject/testdata/auth.yaml.injected (hunk 9 above)
Comment (rkpagadala) [reply platform/kube/inject/testdata/auth.yaml.injected]: changed to text based now the command line is
--controlPlaneAuthPolicy NONE
or
--controlPlaneAuthPolicy MUTUAL_TLS

File: cmd/pilot-agent/main.go [hunk 11]
```diff
@@ -119,6 +125,30 @@ var (
...
 			proxyConfig.ProxyAdminPort = int32(proxyAdminPort)
 
+			var pilotSAN []string
+			switch controlPlaneAuthPolicy {
+			case proxyconfig.AuthenticationPolicy_NONE.String():
+				proxyConfig.ControlPlaneAuthPolicy = proxyconfig.AuthenticationPolicy_NONE
+			case proxyconfig.AuthenticationPolicy_MUTUAL_TLS.String():
+				var ns string
+				proxyConfig.ControlPlaneAuthPolicy = proxyconfig.AuthenticationPolicy_MUTUAL_TLS
+				if serviceregistry == platform.KubernetesRegistry {
```
Comment (wattli) [inline cmd/pilot-agent/main.go]: What about consul/eureka? Do we have plan to support them as well?

File: platform/kube/inject/testdata/auth.cert-dir.yaml.injected [hunk 12]
```diff
@@ -41,7 +41,7 @@ spec:
         - --parentShutdownDuration
         - 3s
         - --discoveryAddress
-        - istio-pilot:8080
```
Comment (wattli) [inline platform/kube/inject/testdata/auth.cert-dir.yaml.injected]: Why make this change throughout the whole PR?

File: cmd/pilot-agent/main.go (hunk 11 above)
Comment (rshriram) [reply cmd/pilot-agent/main.go]: We don't have Istio auth running in those environments.

File: platform/kube/inject/testdata/auth.cert-dir.yaml.injected (hunk 12 above)
Comment (rkpagadala) [reply platform/kube/inject/testdata/auth.cert-dir.yaml.injected]: Upgrade becomes easier, we need separate ports for envoy in front of mixer and pilot.

File: cmd/pilot-agent/main.go [hunk 13]
```diff
@@ -51,6 +52,8 @@ var (
 	connectTimeout         time.Duration
 	statsdUDPAddress       string
 	proxyAdminPort         int
+	controlPlaneAuthPolicy string
```
Comment (costinm) [inline cmd/pilot-agent/main.go]: I still believe this should be called 'allowInsecure', but we can review the API/config naming in a different forum...

File: cmd/pilot-agent/main.go [hunk 14]
```diff
@@ -91,9 +94,11 @@ var (
 					role.ID = role.IPAddress
 				}
 			}
+			pilotDomain := role.Domain
 			if role.Domain == "" {
 				if serviceregistry == platform.KubernetesRegistry {
 					role.Domain = os.Getenv("POD_NAMESPACE") + ".svc.cluster.local"
+					pilotDomain = "cluster.local"
```
Comment (costinm) [inline cmd/pilot-agent/main.go]: TODO: remove hardcoded cluster.local !

File: cmd/pilot-agent/main.go (hunk 11 above)
Comment (costinm) [reply cmd/pilot-agent/main.go]: We may still want to secure them - some script generating equivalent certs ( but maybe longer lived since agent is not there), or using the agent may be possible.

File: cmd/pilot-agent/main.go [hunk 15]
```diff
@@ -221,6 +251,10 @@ func init() {
 		"IP Address and Port of a statsd UDP listener (e.g. 10.75.241.127:9125)")
 	proxyCmd.PersistentFlags().IntVar(&proxyAdminPort, "proxyAdminPort", int(values.ProxyAdminPort),
 		"Port on which Envoy should listen for administrative commands")
+	proxyCmd.PersistentFlags().StringVar(&controlPlaneAuthPolicy, "controlPlaneAuthPolicy",
+		values.ControlPlaneAuthPolicy.String(), "Control Plane Authentication Policy")
```
Comment (costinm) [inline cmd/pilot-agent/main.go]: Can we at least make the comment clear that NONE means no TLS/security for the control plane, and is strongly discouraged ?

File: docker/Dockerfile.proxy [hunk 16]
```diff
@@ -1,4 +1,10 @@
 # Change this and ../WORKSPACE at the same time
 FROM gcr.io/istio-testing/envoy:cf01c6e70aceccbad9f0ecdf4c61f4918ef0d93b
 ADD pilot-agent /usr/local/bin/pilot-agent
+
+COPY envoy_pilot.json      /etc/istio/proxy/envoy_pilot.json
```
Comment (costinm) [inline docker/Dockerfile.proxy]: Please make sure the files are also included in the .deb file in proxy package.

File: docker/start.sh [hunk 17]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+count=0
```
Comment (costinm) [inline docker/start.sh]: - check indentation ? (separate PR is ok)
- this should be in the base image, generated by the proxy package, so it can be reused for 'mesh expansion'.

Most files should be moved there - it is also ok to do that in a separate PR, to unblock the testing, but it must be done.

File: cmd/pilot-agent/main.go (hunk 15 above)
Comment (costinm) [reply cmd/pilot-agent/main.go]: Also - make sure default is 'NONE' and add a comment that default will change to TLS (we need to allow upgrades)

**Comment by rkpagadala** [comment]
/assign @rshriram
please add a /approve

**Comment by rkpagadala** [comment]
@rshriram please add /approve label

File: cmd/pilot-agent/main.go [hunk 18]
```diff
@@ -119,6 +125,30 @@ var (
 			proxyConfig.StatsdUdpAddress = statsdUDPAddress
 			proxyConfig.ProxyAdminPort = int32(proxyAdminPort)
 
+			var pilotSAN []string
```
Comment (ijsnellf) [inline cmd/pilot-agent/main.go]: put this in a function to make `main.go` more readable

Comment (rkpagadala) [reply cmd/pilot-agent/main.go]: this pr got merged, will fix in another PR
